### ISSUE TITLE
Run periodic tests against the minimum supported Kubernetes version

### ISSUE THREAD

**Issue by sedefsavas** [kubernetes-sigs/cluster-api-provider-aws#3087]
We already test the maximum K8s version (have a conformance test that is running against the main K8s branch), we should also do the same for the minimum supported version.

Replicating this from https://github.com/kubernetes-sigs/cluster-api/issues/5963, see it for more details.

**Comment by rayandas** [comment]
@sedefsavas Can I work on this?

**Comment by sedefsavas** [comment]
@pydctw could you follow up on this for pointers if needed?

**Comment by rayandas** [comment]
I need few inputs here:
What would be the job name, testgrid tab name, min supported K8s version that we want to add and the ideal interval?

@pydctw

**Comment by pydctw** [comment]
@sedefsavas, a clarifying question before I give an answer to @rayandas. 

The CAPI issue kubernetes-sigs/cluster-api#5963 is about adding a periodic test to run `unit/integration` test against min k8s version. 
We have periodic e2e tests but no periodic unit/integration test. Is this issue about adding a periodic conformance test or unit/integration test?

**Comment by sedefsavas** [comment]
@pydctw 
I was thinking about adding e2e tests like this: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-azure#capi-periodic-upgrade-main-1-19-1-20&show-stale-tests=

I don't see much benefit of periodically running unit tests as PRs have blocking unit tests, but let's research on this to decide.

**Comment by pydctw** [comment]
Sounds great, @sedefsavas 

@rayandas, as you will be adding a periodic e2e test running against min k8s version, I would pick sth like `periodic-e2e-mink8s-main` as testgrid tab name. Current min k8s version is 1.19. For other parameters, you can decide accordingly.

You can review the [existing test](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api-provider-aws/cluster-api-provider-aws-periodics.yaml#L1-L42) and create a new one with modified parameters.

As another reference, see https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-main.yaml#L144-L182

**Comment by rayandas** [comment]
Alright! thanks. @pydctw :)

**Comment by sedefsavas** [comment]
@rayandas The PR open does not work as is because we do not have special handling implemented like CAPZ.
Image resolution fails when an unpublished image version is provided like stable-1.19
https://github.com/kubernetes/test-infra/pull/25249

`ClusterctlUpgradeTests` would be better to use for this issue as pointed above, and I guess @pydctw would be a good resource as she moved some of our tests recently to use it.

We need to have a custom image resolution similar to what CAPZ has implemented here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1767
Basically, strip out `stable` section from the version name and find the latest published image version. (`	resolveKubernetesVersions(config)`)

We already have a similar logic here `conformanceImageID()`, we need to extend it to use it for this purpose.

And then add upgrade tests like this: https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api-provider-azure/cluster-api-provider-azure-periodics-main-upgrades.yaml

**Comment by sedefsavas** [comment]
@rayandas 
Rethinking about this issue and discussing with @pydctw offline, as per @pydctw's suggestion let's make this much simpler in the first iteration:

We will have this test in test infra:
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api-provider-azure/cluster-api-provider-azure-periodics-main-upgrades.yaml#L3

But instead of passing ["stable-1.19"](https://github.com/kubernetes/test-infra/blob/aeeaba2bd2fbc6d34f9b4a11a7c177a3df8e4a2d/config/jobs/kubernetes-sigs/cluster-api-provider-azure/cluster-api-provider-azure-periodics-main-upgrades.yaml#L27
) and ["stable-1.20"](https://github.com/kubernetes/test-infra/blob/aeeaba2bd2fbc6d34f9b4a11a7c177a3df8e4a2d/config/jobs/kubernetes-sigs/cluster-api-provider-azure/cluster-api-provider-azure-periodics-main-upgrades.yaml#L29) versions we will use exact versions: v1.19.6 to v1.20.3

By this way no custom AMI resolution will be needed. CAPZ did this to avoid updating patch versions.

Feel free to reach out if you have questions.

**Comment by rayandas** [comment]
Ah! Alright. @sedefsavas So the value `GINKGO_FOCUS` will be only `K8s-Upgrade` for the new test right?

**Comment by sedefsavas** [comment]
Yes, just created a PR so that we can use focus to call all cluster upgrade tests from different Prow jobs.
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3220

**Comment by rayandas** [comment]
@sedefsavas I raised the PR, please review it and let me know if any changes are required. I added the test under `cluster-api-provider-aws-periodics.yaml`, if we're going to add other upgrade tests in future, then maybe we can create a new yaml file and keep the upgrade tests there (e.g. cluster-api-provider-azure-periodics-main-upgrades.yaml)
